src/Components/PresentationalComponents/Filters/PrimaryToolbarFilters/AppliesToOSFilter.js

```diff
@@ -0,0 +1,88 @@
+import { conditionalFilterType } from '@redhat-cloud-services/frontend-components/ConditionalFilter';
+import { intl } from '../../../../Utilities/IntlProvider';
+import messages from '../../../../Messages';
+import buildOSGroups from '../../../SmartComponents/Reports/Common/buildOSGroups';
+
+const convertOperatingSystemsToFilterGroupItems = (operatingSystems) => {
+    const groupedByOS = buildOSGroups(operatingSystems);
+    const filterGroupItems = Object.entries(groupedByOS).map(([groupName, items]) => ({
+        label: groupName,
+        value: groupName,
+        name: groupName,
+        id: groupName,
+        type: 'treeView',
+        groupSelectable: true,
+        children: items.map(item => ({
+            label: item.label,
+            value: item.value,
+            name: item.label,
+            id: item.value,
+            groupSelectable: true,
+            type: 'treeView'
+        }))
+    }));
+    return filterGroupItems;
+};
+
+const buildSelectedVersionsObject = (selectedVersionsArray) => {
+    return selectedVersionsArray.reduce((acc, version) => {
+        const majorVersion = version.split('.')[0];
+        const osGroup = `RHEL ${majorVersion}`;
+
+        if (!acc[osGroup]) {
+            acc[osGroup] = {};
+        }
+
+        acc[osGroup][version] = true;
+
+        return acc;
+    }, {});
+};
+
+const appliesToOSFilter = (apply, currentFilter = {}, operatingSystems = []) => {
+    let { rhel_version: selectedVersionsContext = '' } = currentFilter;
+    let selectedVersionsArray = selectedVersionsContext ? selectedVersionsContext.split(',').map(version => version.trim()) : [];
+
+    const filterByOSVersions = values => {
+        let updatedVersions = [...selectedVersionsArray];
+
+        Object.entries(values).forEach((entry) => {
+            const versions = entry[1];
+            Object.keys(versions).forEach(version => {
+                if (versions[version]) {
+                    // If the version is selected, add it to updatedVersions if it's not there yet
+                    if (!updatedVersions.includes(version)) {
+                        updatedVersions.push(version);
+                    }
+                } else {
+                    // If the version is deselected, remove it from updatedVersions
+                    updatedVersions = updatedVersions.filter(v => v !== version);
+                }
+            });
+        });
```

Comment:
I wonder if we can't reuse the code I'm adding here in my PR to do some of the same work?

https://github.com/RedHatInsights/vulnerability-ui/pull/2236/files#diff-f281aa47dd64aab4c195c7e5a6f7e430c3a7dabb17432ca9af03331245af3958R78-R82

Reply:
As we agreed, I'm putting a comment to review/refactor it later after your PR is merged